src/frontend/src/optimizer/plan_visitor/cardinality_visitor.rs

```diff
@@ -0,0 +1,402 @@
+            Self::new(max(self.lo(), rhs.lo()), max(self.hi, rhs.hi))
+        }
+
+        /// Returns the cardinality with both lower and upper bounds limited to `limit`.
+        pub fn limit_to(self, limit: usize) -> Self {
+            self.min(Self::exact(limit))
+        }
+
+        /// Returns the cardinality with the lower bound limited to `limit`.
+        pub fn as_low_as(self, limit: usize) -> Self {
+            self.min(Self::at_least(limit))
+        }
+    }
+
+    impl Add<Cardinality> for Cardinality {
+        type Output = Self;
+
+        /// Returns the sum of the two cardinalities, where the lower and upper bounds are
+        /// respectively the sum of the lower and upper bounds of the two cardinalities.
+        fn add(self, rhs: Self) -> Self::Output {
+            let lo = self.lo().saturating_add(rhs.lo());
+            let hi = if let (Some(lhs), Some(rhs)) = (self.hi(), rhs.hi()) {
+                lhs.checked_add(rhs)
+            } else {
+                None
+            };
+            Self::new(lo, hi)
+        }
+    }
+
+    impl Sub<usize> for Cardinality {
+        type Output = Self;
+
+        /// Returns the cardinality with both lower and upper bounds subtracted by `rhs`.
+        fn sub(self, rhs: usize) -> Self::Output {
+            let lo = self.lo().saturating_sub(rhs);
+            let hi = self.hi().map(|hi| hi.saturating_sub(rhs));
+            Self::new(lo, hi)
+        }
+    }
+
+    impl Mul<Cardinality> for Cardinality {
+        type Output = Cardinality;
+
+        /// Returns the product of the two cardinalities, where the lower and upper bounds are
+        /// respectively the product of the lower and upper bounds of the two cardinalities.
+        fn mul(self, rhs: Cardinality) -> Self::Output {
+            let lo = self.lo().saturating_mul(rhs.lo());
+            let hi = if let (Some(lhs), Some(rhs)) = (self.hi(), rhs.hi()) {
+                lhs.checked_mul(rhs)
+            } else {
+                None
+            };
+            Self::new(lo, hi)
+        }
+    }
+
+    impl Mul<usize> for Cardinality {
+        type Output = Self;
+
+        /// Returns the cardinality with both lower and upper bounds multiplied by `rhs`.
+        fn mul(self, rhs: usize) -> Self::Output {
+            let lo = self.lo().saturating_mul(rhs);
+            let hi = self.hi().and_then(|hi| hi.checked_mul(rhs));
+            Self::new(lo, hi)
+        }
+    }
+
+    impl Cardinality {
+        /// Returns the cardinality if it is exact, `None` otherwise.
+        pub fn get_exact(self) -> Option<usize> {
+            self.hi().filter(|hi| *hi == self.lo())
+        }
+
+        /// Returns `true` if the cardinality is at most `count` rows.
+        pub fn is_at_most(self, count: usize) -> bool {
+            self.hi().is_some_and(|hi| hi <= count)
+        }
+    }
+
+    #[cfg(test)]
+    mod tests {
+        use super::*;
+
+        #[test]
+        fn test_limit_to() {
```

Comment:
What about using doctests? It's quite hard to know the meaning of the methods without looking at examples.

Reply:
Good point. I'll move them.